cmd/rocsp-tool/main.go

```diff
@@ -118,31 +157,364 @@ func main2() error {
+
+func (i *inflight) remove(n uint64) {
+	i.Lock()
+	defer i.Unlock()
+	delete(i.items, n)
+}
+
+func (i *inflight) len() int {
+	i.RLock()
+	defer i.RUnlock()
+	return len(i.items)
+}
+
+// min returns the numerically smallest key inflight. If nothing is inflight,
+// it returns 0. Note: this takes O(n) time in the number of keys and should
+// be called rarely.
+func (i *inflight) min() uint64 {
+	i.RLock()
+	defer i.RUnlock()
+	if len(i.items) == 0 {
+		return 0
+	}
+	var min uint64
+	for k := range i.items {
+		if min == 0 {
+			min = k
+		}
+		if k < min {
+			min = k
+		}
+	}
+	return min
+}
+
+// processResult represents the result of attempting to sign and store status
+// for a single certificateStatus ID. If `err` is non-nil, it indicates the
+// attempt failed.
+type processResult struct {
+	id  uint64
+	err error
+}
+
+func (cl *client) loadFromDB(ctx context.Context, speed ProcessingSpeed) error {
+	// To scan the DB efficiently, we want to select only currently-valid certificates. There's a
+	// handy expires index, but for selecting a large set of rows, using the primary key will be
+	// more efficient. So first we find a good id to start with, then scan from there. Note: since
+	// AUTO_INCREMENT can skip around a bit, we add padding to ensure we get all currently-valid
+	// certificates.
+	startTime := cl.clk.Now().Add(-24 * time.Hour)
+	var minID *int64
+	err := cl.db.QueryRowContext(
+		ctx,
+		"SELECT MIN(id) FROM certificateStatus WHERE notAfter >= ?",
+		startTime,
+	).Scan(&minID)
+	if err != nil {
+		return fmt.Errorf("selecting minID: %w", err)
+	}
+	if minID == nil {
+		return fmt.Errorf("no entries in certificateStatus (where notAfter >= %s)", startTime)
+	}
+
+	// Limit the rate of reading rows.
+	frequency := time.Duration(float64(time.Second) / float64(time.Duration(speed.RowsPerSecond)))
+	// a set of all inflight certificate statuses, indexed by their `ID`.
+	inflightIDs := newInflight()
+	statusesToSign := cl.scanFromDB(ctx, *minID, frequency, inflightIDs)
+
+	results := make(chan processResult, speed.ParallelSigns)
+	var runningSigners int32
+	for i := 0; i < speed.ParallelSigns; i++ {
+		atomic.AddInt32(&runningSigners, 1)
+		go cl.signAndStoreResponses(ctx, statusesToSign, results, &runningSigners)
+	}
+
+	var successCount, errorCount int64
+
+	for result := range results {
+		inflightIDs.remove(result.id)
+		if result.err != nil {
+			errorCount++
+			if errorCount < 10 ||
+				(errorCount < 1000 && rand.Intn(1000) < 100) ||
+				(errorCount < 100000 && rand.Intn(1000) < 10) ||
+				(rand.Intn(1000) < 1) {
+				log.Printf("error: %s", result.err)
+			}
+		} else {
+			successCount++
+		}
+
+		if (successCount+errorCount)%10 == 0 {
+			log.Printf("stored %d responses, %d errors", successCount, errorCount)
+		}
+	}
+
+	log.Printf("done. processed %d successes and %d errors\n", successCount, errorCount)
+	if inflightIDs.len() != 0 {
+		return fmt.Errorf("inflightIDs non-empty! has %d items", inflightIDs.len())
```

Comment:
Why not print the lowest inflight ID here as well? Sure printing it every 10k lines is good in case this crashes, but might as well output it at the last possible moment too, for the "normal" case.